prusti-encoder/src/encoders/typ.rs

```diff
@@ -200,6 +229,385 @@ impl TaskEncoder for TypeEncoder {
                 None
             )
         }
+
+        fn mk_enum<'tcx, 'vir>(
+            vcx: &'vir vir::VirCtxt<'tcx>,
+            deps: &mut TaskEncoderDependencies<'vir>,
+            adt: ty::AdtDef<'tcx>,
+            task_key: &<TypeEncoder as TaskEncoder>::TaskKey<'tcx>,
+        ) -> Result<
+            <TypeEncoder as TaskEncoder>::OutputFullLocal<'vir>,
+            (
+                <TypeEncoder as TaskEncoder>::EncodingError,
+                Option<<TypeEncoder as TaskEncoder>::OutputFullDependency<'vir>>,
+            ),
+        > {
+            let did_name = vcx.tcx.item_name(adt.did()).to_ident_string();
+
+            let name_s = vir::vir_format!(vcx, "s_Adt_{did_name}");
+            let name_p = vir::vir_format!(vcx, "p_Adt_{did_name}");
+
+            let field_discriminant = vir::vir_format!(vcx, "p_Adt_{did_name}_discriminant");
+
+            let ty_s = vcx.alloc(vir::TypeData::Domain(name_s));
+
+            let s_discr_func_name = vir::vir_format!(vcx, "{name_s}_discriminant");
+           
+            let discr_bit_width = {
+                use crate::rustc_middle::ty::util::IntTypeExt;
+                TypeEncoder::get_bit_width(vcx.tcx, adt.repr().discr_type().to_ty(vcx.tcx))
+            };
+
+            let discr_type = vcx.alloc(vir::TypeData::Int {
```

Comment:
@tillarnold you should definitely not be manually creating a new type here! Why do we even have a `TypeEncoder`... you have a `ty::Ty` from `adt.repr().discr_type().to_ty(vcx.tcx)` make sure to recurse and pass that in to the type encoder, this way you'd avoid code duplications multiple bugs:
- Hardcoding `signed: false`
- Using `TypeData::Int` rather than the snapshot version

Note that the point the Viper native `TypeData::Int` is _only_ for use with Unary and Binary operators and any other place where Viper explicitly needs native Ints. __None of the Rust integer types can be correctly represented by `Int`__: an `Int` is an unbounded value while all Rust values are bounded.

Reply:
Just fyi for the future: I will fix this now

Reply:
That makes a lot of sense. I had wrongly assumed that we don't need the discriminant to be a rust type, as e.g. [std::mem::Discriminant](https://doc.rust-lang.org/std/mem/struct.Discriminant.html#) doesn't give access to the primtive type. But that is obviously not true e.g. with `repr`, `as` and with things on the MIR level. Sorry about that.